### ISSUE TITLE
Finetuning MWP with custom masks?

### ISSUE THREAD

**Issue by yuchenlin** [EricFillion/happy-transformer#125]
Hi Eric,

Thanks for the feature of fine-tuning MWP. I am wondering if we can write our own customized masking strategy instead of random sampling?  

Simply put, I have a list of sentences where masks are already created, and we know the associated words for each mask. Can we fine-tune HappyBERT/HappyROBERTA for them? 

Thanks!

**Comment by yuchenlin** [comment]
It seems that I can modify the code [here](https://github.com/EricFillion/happy-transformer/blob/e42abe97fb20f247f8548a4b4bb781ef4297c2c3/happytransformer/mlm_utils.py#L73)

**Comment by uGokalp** [comment]
Let me see if I understand your question. You want to be able to use your own "mask_tokens" function. If your function takes uses the same arguments and outputs inputs, labels like "mask_tokens"  I should be able to accommodate your request.

**Comment by yuchenlin** [comment]
Hi Umur, thanks for the quick reply. Almost! What I have now is a list of tuples as follows, where each tuple is a masked sentence and the original word being masked.

```
[ ("I have a  <mask> dog. ", "cute"),
  ("You have a  <mask> cat. ", "big"),
  ("He has a  <mask> girlfriend. ", "pretty"),
 ....
]
```
These results are created by my custom word-masking script. 

Thus, the masks and the original words are already created before I wanna call the `mask_tokens` function so that the argument can be a little bit different.

**Comment by uGokalp** [comment]
Okay, I think I understand now. I'll get started on this and let you know when you can test it.

**Comment by yuchenlin** [comment]
Thanks!

**Comment by Rahul-Khanna** [comment]
hey Eric, so I kinda implemented a "hacky" way of doing this in my fork

https://github.com/Rahul-Khanna/happy-transformer/blob/master/happytransformer/mlm_utils.py#L114

its a bit hacky as the tokenizer could technically split one of the words a person is looking to mask, but then I'm not even sure what the user would want in that situation.  Also of course the word could match to a broken up word or other instance of the word, but in my use case this isn't going to happen. https://github.com/Rahul-Khanna/happy-transformer/blob/master/happytransformer/mlm_utils.py#L143

Also I'm not handling the case of a batch_size > 1

Here's how one would use it:
https://github.com/Rahul-Khanna/happy-transformer/blob/master/examples/train_happyroberta.py

The data files can be found here: https://github.com/Rahul-Khanna/common_sense_embedding_analysis/tree/rahul/data/finetune_data

Just thought I'd let you and @yuchenlin know

**Comment by EricFillion** [comment]
@Rahul-Khanna 

Thank you so much for your contribution! We'll look into applying your code for our next release and give you credit for it.

**Comment by Rahul-Khanna** [comment]
yup! I'm going to keep adding to my fork, but happy to create a pr here when I have something more stable.

**Comment by EricFillion** [comment]
@Rahul-Khanna 
Sounds great, I'm looking forward to it.
